tooling/lsp/src/requests/inlay_hint.rs

```diff
@@ -0,0 +1,560 @@
+use std::future::{self, Future};
+
+use async_lsp::ResponseError;
+use fm::{FileId, PathString};
+use lsp_types::{
+    InlayHint, InlayHintKind, InlayHintLabel, InlayHintLabelPart, InlayHintParams, Position,
+    TextDocumentPositionParams,
+};
+use noirc_errors::{Location, Span};
+use noirc_frontend::{
+    self,
+    ast::{
+        BlockExpression, Expression, ExpressionKind, LetStatement, NoirFunction, Pattern,
+        Statement, StatementKind, TraitImplItem, TraitItem, UnresolvedTypeData,
+    },
+    node_interner::ReferenceId,
+    parser::{Item, ItemKind},
+    ParsedModule, Type, TypeBinding, TypeVariable, TypeVariableKind,
+};
+
+use crate::LspState;
+
+use super::{process_request, to_lsp_location, ProcessRequestCallbackArgs};
+
+pub(crate) fn on_inlay_hint_request(
+    state: &mut LspState,
+    params: InlayHintParams,
+) -> impl Future<Output = Result<Option<Vec<InlayHint>>, ResponseError>> {
+    let text_document_position_params = TextDocumentPositionParams {
+        text_document: params.text_document.clone(),
+        position: Position { line: 0, character: 0 },
+    };
+
+    let result = process_request(state, text_document_position_params, |args| {
+        let path = PathString::from_path(params.text_document.uri.to_file_path().unwrap());
+        args.files.get_file_id(&path).map(|file_id| {
+            let file = args.files.get_file(file_id).unwrap();
+            let source = file.source();
+            let (parsed_moduled, _errors) = noirc_frontend::parse_program(source);
+
+            // The version of codespan_lsp is pretty old and relies on an old version of lsp-types.
+            // We can't downgrade because we need some types from the latest version, and there's
+            // no newer version of codespan_lsp... but we just need this type here, so we create
+            // a copy of a Range to get an older type.
+            let range = lsp_types_0_88_0::Range {
+                start: lsp_types_0_88_0::Position {
+                    line: params.range.start.line,
+                    character: params.range.start.character,
+                },
+                end: lsp_types_0_88_0::Position {
+                    line: params.range.end.line,
+                    character: params.range.end.character,
+                },
+            };
+
+            let span = codespan_lsp::range_to_byte_span(args.files, file_id, &range)
+                .ok()
+                .map(|range| Span::from(range.start as u32..range.end as u32));
+
+            let mut collector = InlayHintCollector::new(args, file_id, span);
+            collector.collect_in_parsed_module(&parsed_moduled);
+            collector.inlay_hints
+        })
+    });
+    future::ready(result)
+}
+
+pub(crate) struct InlayHintCollector<'a> {
+    args: ProcessRequestCallbackArgs<'a>,
```

Comment:
We only use `interner` and `files` from this right? Can we pull out these fields from `args` so it's clearer on what we need?